[Step3] 자동차 경주 게임

chomily opened the pull request:
안녕하세요!
자동차 경주 게임 Step3 미션 제출합니다 
리뷰 부탁드려요~😄

chomily committed:
refactor based on feedback

- narrow down Exception to catch in try-catch statement
- add return type at function
- change class name Number -> Operand
- remove unused function
- use method chaining

chomily committed:
refactor based on feedback

- remove StringCalculator.expression property
- make StringCalculator#calculator static function

chomily committed:
feat: add InputView, ResultView

chomily committed:
feat: add Car#move

chomily committed:
feat: add NumberGenerator interface, RandomIntGenerator class

chomily committed:
refactor: move packages

chomily committed:
refactor: Car#move accept NumberGenerator as parameter, add Movement Enum

chomily committed:
feat: car can save movement histories

chomily committed:
refactor: extract MoveRule Interface

chomily committed:
feat: add RacingGame, Application

chomily committed:
refactor: make Car#move void fun and bug fix at printing result

malibinYun commented on src/main/kotlin/com/nextstep/racingcar/domain/Car.kt:
```diff
@@ -0,0 +1,19 @@
+package com.nextstep.racingcar.domain
+
+class Car {
+    private val histories = mutableListOf<Movement>()
```
자동차가 기록을 가지게 하기 보다 움직인 거리만 가지고, 외부에서 기록을 하게 만들어보는 것은 어떨까요?
혹은, 기록을 하는 객체를 따로 만들어보는 것은 어떨까요?

chomily replied:
그러게요. History 객체를 따로 만들면 레이싱 카 게임에 계정을 적용하거나 다른 기능들을 추가하기 용의하겠네요!
외부에 기록을 만들지는 않고 Car 내부에서 history 객체를 가지도록 리팩토링했습니다~

malibinYun commented on src/main/kotlin/com/nextstep/racingcar/domain/Car.kt:
```diff
@@ -0,0 +1,19 @@
+        saveHistory(movement)
+    }
+
+    fun saveHistory(movement: Movement) {
+        histories.add(movement)
+    }
```
test만을 위해 메서드를 외부로 공개하는 것은 지양하는 게 좋아요 :)
외부에 공개하지 않아도 될 것을 공개함으로써 캡슐화가 깨지기 때문이에요.

chomily replied:
그렇네요 반영하겠습니다 :)

malibinYun commented on src/main/kotlin/com/nextstep/racingcar/domain/NumberMoveRule.kt:
```diff
@@ -0,0 +1,11 @@
+
+private const val MOVE_CONDITION = 4
+
+class NumberMoveRule : MoveRule {
+
+    override fun move(number: Int): Movement {
+        require(number in 0..9) { "Number should be greater than or equal 0 and less than or equal 9" }
+        return if (number >= MOVE_CONDITION) Movement.MOVE else Movement.NONE
+    }
+}
```
MoveRule에 NumberGenerator의 역할이 함께 있어도 괜찮다는 생각이 드네요 :)

malibinYun commented on src/main/kotlin/com/nextstep/racingcar/domain/RacingGame.kt:
```diff
@@ -0,0 +1,32 @@
+import com.nextstep.racingcar.view.InputView
+import com.nextstep.racingcar.view.ResultView
+
+class RacingGame(private val numberGenerator: NumberGenerator, private val moveRule: MoveRule) {
+    private val inputView = InputView()
+    private val resultView = ResultView()
```
view들도 생성자로부터 받아오는 것은 어떨까요?
어떤 장점들이 생길 수 있을까요?

chomily replied:
아하 좋으네요~ view 내용이 바뀌어도 RacingGame 클래스 코드의 변경은 막을 수 있겠군요~!

malibinYun commented on src/main/kotlin/com/nextstep/racingcar/domain/RacingGame.kt:
```diff
@@ -0,0 +1,32 @@
+        val numberOfCars = inputView.inputNumberOfCars()
+        val numberOfRounds = inputView.inputNumberOfRounds()
+
+        val cars = mutableListOf<Car>()
+        for (i in 1..numberOfCars) {
+            cars.add(Car())
+        }
+
```
```suggestion
        val cars = (1..numberOfCars).map { Car() }
```
이렇게 코틀린의 확장함수를 활용해볼 수 있겠네요 :)

malibinYun commented on src/main/kotlin/com/nextstep/racingcar/domain/RacingGame.kt:
```diff
@@ -0,0 +1,32 @@
+        }
+
+        resultView.printResult()
+        for (round in 1..numberOfRounds) {
+            play(cars)
+        }
```
repeat() 를 활용해보는 것은 어떨까요?
다른 for문들도 다른 확장함수로 바꿔보는 것을 추천드려요 :)

chomily replied:
오호 repeat라는 문(?)이 있군요! 감사합니다~ 수정했어요!

malibinYun commented on src/main/kotlin/com/nextstep/stringcalculator/StringCalculator.kt:
```diff
@@ -1,11 +1,11 @@
 
-    fun calculate(): Long = expression.calculate()
+    companion object {
+        fun calculate(input: String?): Long {
+            require(input != null && input.isNotBlank()) { "Input shouldn't be null or blank" }
```
String?.isNullOrBlank()를 활용해볼 수 있겠네요 :)
```suggestion
        fun calculate(input: String?): Long {
            require(input.isNullOrBlank()) { "Input shouldn't be null or blank" }
```

malibinYun commented on src/test/kotlin/com/nextstep/racingcar/domain/CarTest.kt:
```diff
@@ -0,0 +1,33 @@
+import com.nextstep.racingcar.domain.Movement.NONE
+import io.kotest.core.spec.style.StringSpec
+import io.kotest.matchers.collections.shouldContainExactly
+import io.mockk.every
+import io.mockk.mockk
+import io.mockk.verify
+
+class CarTest : StringSpec({
+    val car = Car()
+    val numberGenerator = mockk<NumberGenerator>()
+    val moveRule = mockk<NumberMoveRule>()
+
+    "Car#move returns Movement Enum meaning thee car moved or not" {
+        every { numberGenerator.generate() } returns 5
+        every { moveRule.move(5) } returns MOVE
+
+        car.move(numberGenerator, moveRule)
+
+        verify { numberGenerator.generate() }
+        verify { moveRule.move(any()) }
+    }
```
이번 미션에서는 mock 라이브러리를 활용하지 않고 테스트를 작성해보는 것은 어떨까요?

chomily replied:
어떤 숫자가 생성되었는지에 따라 차가 움직이는데요~
따라서 move 내부에서 의존하고 있는 객체가 리턴해주는 값을 목으로 주입해주었습니다~

혹시 mock을 사용하지 않았으면 하는 이유가 무엇일까요~?

malibinYun replied:
오늘 강좌에서 배우신 것 처럼, 랜덤 요소 등을 외부 주입을 활용해 테스트할 수 있게 바꿀 수 있었어요.
이 부분들은 모두 mock 라이브러리의 도움 없이 테스트가 가능합니다 :)

또, mock 라이브러리를 사용하면, 외부 주입을 받지 않고도 코드 작성을 할 수 있기 때문에,
이러한 요소들의 테스트 연습을 위해서 mock 테스트 라이브러리 사용을 권장하지 않고 있어요 :)

malibinYun reviewed (changes requested):
3단계 미션 정말 고생 많으셨어요!!
생각해보면 좋을 점들에 코멘트 남겨두었으니 확인 부탁드릴게요!
피드백 반영하시고, 다시 리뷰 요청바랍니다 :)
반이나 미션 진행이 되었네요! 완주까지 화이팅이에요 💪

chomily committed:
refactor based on feedbacks

- utilize Kotlin APIs
- inverse objects' dependency(RacingGame, InputView, ResultView)

chomily committed:
refactor based on feedbacks

- NumberMoveRule has RandomNumberGenerator
- make saveHistory fun private which isn't used except test codes
- reorganize packages

chomily committed:
refactor based on feedbacks

- extract MoveHistory class
- rename package

malibinYun commented on src/main/kotlin/com/nextstep/racingcar/domain/RacingGame.kt:
```diff
@@ -0,0 +1,31 @@
+
+class RacingGame(
+    private val moveRule: MoveRule,
+    private val inputView: InputView,
+    private val resultView: CliResultView
```
View를 interface를 활용해서 구현해주셨네요 👍
ResultView쪽은 실수이실까요 ?_?

chomily replied:
어잌쿠 그러네요...! 수정했습니다 :)

malibinYun commented on src/main/kotlin/com/nextstep/racingcar/domain/RacingGame.kt:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    private fun play(cars: List<Car>) {
+        cars.map { car ->
+            car.move(moveRule)
+            resultView.printResult(car)
+        }
```
.map{} 으로 반환되는 List를 사용하지 않으니, repeat() 혹은, .forEach()를 활용해보는 것은 어떨까요?

chomily replied:
오호, 넵! 자바 스트림을 생각하고 map, forEach 등을 사용하는 게 비용이 클 것이라고 생각했었는데요~ 타고 들어가보니 내부적으로 for 문을 사용하는 인라인 함수들이군요,,,! forEach로 수정해보았습니다🙏

malibinYun replied:
말씀해주신 것 처럼, 코틀린 collections 확장 함수들을 사용하면, 매 라인 마다 새로운 Collection 객체가 생성되는 것이 단점일 수 있는데요!
이럴 때는 .asSequance() 를 사용하면 됩니다. Java Stream과 거의 비슷한 개념이라고 보시면 되어요 :)

malibinYun commented on src/main/kotlin/com/nextstep/racingcar/domain/rules/NumberMoveRule.kt:
```diff
@@ -0,0 +1,14 @@
+
+private const val MOVE_CONDITION = 4
+
+class NumberMoveRule(private val generator: NumberGenerator) : MoveRule {
+
+    override fun move(): Movement {
+        val number = generator.generate()
+        require(number in 0..9) { "Number should be greater than or equal 0 and less than or equal 9" }
+        return if (number >= MOVE_CONDITION) Movement.MOVE else Movement.NONE
+    }
+}
```
음!! 제가 의도한 것은, MoveRule 하위 구현체로, RandomMoveRule 정도가 적당하다는 의견이었어요.
내부에는 0~9 숫자를 생성해서 그것이 4이상히면 MOVE를 반환 하는 형태를 생각했어요.

랜덤한 움직임 전략이 아닌 구현체를 만든다면, FixedMoveRule 같은 구현체를 만들어서 원하는 Movement를 반환하는 형태를 생각했어요 :)
이런 형태로 NumberGenerator와 MoveRule의 역할이 함께 있어도 괜찮아보인다는 의미였습니다.
만약 NumberGenerator와 MoveRule이 둘 다 존재한다면, 이전 형태의 분리가 더 나아보입니다 :)

chomily replied:
음, 처음에 MoveRule, NumberGenerator를 분리했던 이유는 두 조합으로 다양한 움직임을 만들어 낼 수 있겠다고 생각했기 때문이에요.
동일한 이유로 Rule이 Generator를 주입받도록 작성했었는데요.

리뷰어님 이야기를 듣고 지금 보니 이 방식이 오히려 사용성이 떨어질 것 같네요! MoveRule이 항상 `숫자를 생성받아서 그 숫자로 움직임을 판단한다`는 조건에 갇혀버릴 것 같아요. 말씀하신데로 수정해보겠습니다 :)

chomily replied:
다만,,,,, 숫자 생성하는 부분을 private 함수로 뺐더니 테스트를 어떻게 해야할 지 난관에 부딪혔습니다 ㅠㅠ..
```
class RandomMoveRule : MoveRule {

    override fun move(): Movement {
        val number = generate()
        require(number in 0..9) { "Number should be greater than or equal 0 and less than or equal 9" }
        return if (number >= MOVE_CONDITION) Movement.MOVE else Movement.NONE
    }

    private fun generate(): Int {
        val range = 0..9
        return range.random()
    }
}

malibinYun reviewed (approved):
3단계 미션 고생 많으셨습니다 :)
몇 가지 코멘트 남겨두었으니, 확인 해주시고 다음 미션에 반영해주세요!
다음 미션 진행을 위해 머지하겠습니다~

malibinYun commented:
MoveHistory, RacingGame 등의 클래스들도 모두 단위 테스트를 작성 부탁드릴게요~! 🙂
